python/pyspark/context.py

```diff
@@ -1023,6 +1032,41 @@ def getConf(self):
         conf.setAll(self._conf.getAll())
         return conf
 
+    def install_packages(self, packages):
+        """
+        Install python packages on all executors and driver through pip. pip will be installed
+        by default no matter using native virtualenv or conda. So it is guaranteed that pip is
+        available if virtualenv is enabled.
+        :param packages: string for single package or a list of string for multiple packages
+        """
+        if self._conf.get("spark.pyspark.virtualenv.enabled") != "true":
+            raise RuntimeError("install_packages can only use called when "
+                               "spark.pyspark.virtualenv.enabled is set as true")
+        if isinstance(packages, basestring):
+            packages = [packages]
+        # seems statusTracker.getExecutorInfos() will return driver + exeuctors, so -1 here.
+        num_executors = len(self._jsc.sc().statusTracker().getExecutorInfos()) - 1
+        dummyRDD = self.parallelize(range(num_executors), num_executors)
+
+        def _run_pip(packages, iterator):
+            import pip
+            return pip.main(["install"] + packages)
+
+        # install package on driver first. if installation succeeded, continue the installation
+        # on executors, otherwise return directly.
+        if _run_pip(packages, None) != 0:
+            return
```

Comment:
So wait will this fail silently or how will the user know its failed?

Reply:
The pip install output will be displayed.